auth-cas/cas.php

```diff
@@ -113,10 +116,54 @@ public function getName() {
     return $_SESSION[':cas']['name'];
   }
 
+  private function setCustomFields() {
+    $custom = array();
+    if($this->config->get('cas-custom-attributes') !== null) {
+      // parse rows by splitting newlines, then map to columns
+      $attrs = array_map('str_getcsv', str_getcsv($this->config->get('cas-custom-attributes'), "\n"));
+      // iterate over rows of ["claim-attribute", "form-field"]
+      foreach ($attrs as &$x) {
+        // XXX: silently ignore incorrect column size and missing attrs
+        // XXX: does not disallow reserved fields user, email, name
+        if(count($x) == 2 && phpCAS::hasAttribute(trim($x[0]))) {
+          $custom[trim($x[1])] = phpCAS::getAttribute(trim($x[0]));
+        }
+      }
+    }
+    $_SESSION[':cas']['custom'] = json_encode($custom);
+  }
+
+  public function getCustomFields() {
+    return json_decode($_SESSION[':cas']['custom'], true, 2, JSON_THROW_ON_ERROR);
+  }
+
   public function getProfile() {
-    return array(
-      'email' => $this->getEmail(),
-      'name' => $this->getName());
+    return array_merge(
+      array(
+        'email' => $this->getEmail(),
+        'name' => $this->getName()
+      ), $this->getCustomFields());
+  }
+
+  public function getUsernameField() {
+    $field = $this->config->get('cas-username-form-field');
+    if($field === null) {
+      $field = 'email';
+    }
+    return $field;
+  }
+
+  // NOTE: Username in this context is osTicket username, not CAS user;
+  //       see getUser() for CAS user
+  public function getUsername() {
+    // NOTE: since this value is not cached, avoid calling too much
+    $info = $this->getProfile();
+    $info['user'] = $this->getUser();
```

Comment:
This could potentially return an e-mail address. Could that be a problem if the plugin is configured to login with a username?

Reply:
It's intentionally user-configurable so that if they want to use email as username, they can. It will return email address if the person configuring it wants it to.